core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,153 @@
+    private final ESLogger logger = Loggers.getLogger(IndexFolderUpgrader.class);
+    private final MetaDataStateFormat<IndexMetaData> indexStateFormat = readOnlyIndexMetaDataStateFormat();
+
+    /**
+     * Creates a new upgrader instance
+     * @param settings node settings
+     * @param nodeEnv the node env to operate on
+     */
+    IndexFolderUpgrader(Settings settings, NodeEnvironment nodeEnv) {
+        this.settings = settings;
+        this.nodeEnv = nodeEnv;
+    }
+
+    /**
+     * Moves the index folder found in <code>source</code> to <code>target</code>
+     */
+    void upgrade(final Index index, final Path source, final Path target) throws IOException {
+        boolean success = false;
+        try {
+            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
+            success = true;
+        } catch (NoSuchFileException | FileNotFoundException exception) {
+            // thrown when the source is non-existent because the folder was renamed
+            // by another node (shared FS) after we checked if the target exists
+            logger.error("multiple nodes trying to upgrade [{}] in parallel, retry upgrading with single node",
+                exception, target);
+            throw exception;
+        } finally {
+            if (success) {
+                logger.info("{} moved from [{}] to [{}]", index, source, target);
+                logger.trace("{} syncing directory [{}]", index, target);
+                IOUtils.fsync(target, true);
+            }
+        }
+    }
+
+    /**
+     * Renames <code>indexFolderName</code> index folders found in node paths and custom path
+     * iff {@link #needsUpgrade(Index, String)} is true.
+     * Index folder in custom paths are renamed first followed by index folders in each node path.
+     */
+    void upgrade(final String indexFolderName) throws IOException {
+        for (NodeEnvironment.NodePath nodePath : nodeEnv.nodePaths()) {
+            final Path indexFolderPath = nodePath.indicesPath.resolve(indexFolderName);
+            final IndexMetaData indexMetaData = indexStateFormat.loadLatestState(logger, indexFolderPath);
+            if (indexMetaData != null) {
+                final Index index = indexMetaData.getIndex();
+                if (needsUpgrade(index, indexFolderName)) {
+                    logger.info("{} upgrading [{}] to new naming convention", index, indexFolderPath);
+                    final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings);
+                    if (indexSettings.hasCustomDataPath()) {
+                        // we rename index folder in custom path before renaming them in any node path
+                        // to have the index state under a not-yet-upgraded index folder, which we use to
+                        // continue renaming after a incomplete upgrade.
+                        final Path customLocationSource = nodeEnv.resolveBaseCustomLocation(indexSettings)
```

Comment:
Yeah, I think if the settings are there but the index has not been opened yet (don't ask, it's a weird but valid use case), the location that the settings point to may not exist